src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp

```diff
@@ -0,0 +1,184 @@
+        return com_array<ITerminalConnectionFactory>(factories);
+    }
+
+private:
+    const std::vector<ITerminalConnectionProvider> _providers;
+};
+
+struct WrappedModuleProvider : implements<WrappedModuleProvider, ITerminalConnectionProvider>
+{
+    WrappedModuleProvider(ITerminalConnectionProvider other, HMODULE dll) :
+        _other(other),
+        _dll(dll)
+    {
+    }
+
+    ~WrappedModuleProvider()
+    {
+        // Set stored provider to null before freeing the library so the IUnknown et al destructor implementations can be called
+        _other = nullptr;
+        FreeLibrary(_dll);
+        _dll = 0;
+    }
+
+    ITerminalConnectionFactory GetFactory(guid id)
+    {
+        return _other.GetFactory(id);
+    }
+
+    com_array<ITerminalConnectionFactory> GetFactories()
+    {
+        return _other.GetFactories();
+    }
+
+private:
+    ITerminalConnectionProvider _other;
+    HMODULE _dll;
+};
+
+struct OptionalPackageTerminalConnectionProvider : CompositeTerminalConnectionProvider
+{
+    OptionalPackageTerminalConnectionProvider() :
+        CompositeTerminalConnectionProvider(LoadProvidersFromOptionalPackages())
+    {
+    }
+
+private:
+    static bool endsWith(const std::wstring_view& str, const std::wstring_view& suffix)
+    {
+        return str.size() >= suffix.size() && 0 == str.compare(str.size() - suffix.size(), suffix.size(), suffix);
+    }
+
+    static bool startsWith(const std::wstring_view& str, const std::wstring_view& prefix)
+    {
+        return str.size() >= prefix.size() && 0 == str.compare(0, prefix.size(), prefix);
+    }
+
+    static bool IsPlugin(std::wstring_view path)
+    {
+        return startsWith(path, L"Terminal.Plugin.") && endsWith(path, L".dll");
+    }
+
+    typedef winrt::Microsoft::Terminal::TerminalConnection::ITerminalConnectionProvider(__stdcall* GetConnectionProvider)();
+
+    static void DebugOut(const TCHAR* fmt, ...)
+    {
+        TCHAR s[1025];
+        va_list args;
+        va_start(args, fmt);
+        wvsprintf(s, fmt, args);
+        va_end(args);
+        OutputDebugString(s);
+    }
+
+    static std::vector<ITerminalConnectionProvider> LoadProvidersFromOptionalPackages()
+    {
+        std::vector<ITerminalConnectionProvider> factories;
+
+        auto currentAppPackage = winrt::Windows::ApplicationModel::Package::Current();
+
+        for (auto package : currentAppPackage.Dependencies())
+        {
+            if (package.IsOptional())
+            {
+                auto query = package.InstalledLocation().CreateFileQuery();
+                auto files = query.GetFilesAsync().get();
+
+                for (auto file : files)
+                {
+                    if (!file.IsAvailable())
+                    {
+                        continue;
+                    }
+
+                    auto name = file.Name();
+
+                    if (IsPlugin(name))
+                    {
+                        auto dll = LoadPackagedLibrary(name.data(), 0);

```

Comment:
This is incredible that it was this easy

Reply:
I mean, there's a lot of other stuff going on that's not easy, but I was under the impression that this specific bit would be the hard part. The rest is really just plumbing to make this bit work like magic.